internal/static/cartogram2.js

```diff
@@ -3326,190 +3401,226 @@ class Cartogram {
         {
             if(document.getElementById('csv').files.length < 1)
             {
-                this.doNonFatalError(Error('You must upload CSV data.'));
+                this.doNonFatalError(Error('You must upload CSV/Excel data.'));
                 return false;
             }
+            
+            // convert given xlsx file to csv
+            var input_data_file = document.getElementById('csv').files[0];
+            if(input_data_file.name.split('.').pop() === 'xlsx') {
+                
+                var reader = new FileReader();
+                reader.onloadend = function(e) {
+                    var data = e.target.result;
+                    var wb = XLSX.read(data, {type: 'binary'});
+                    var ws = wb.Sheets[wb.SheetNames[0]];
+                    var csv = XLSX.utils.sheet_to_csv(ws);
+                    csv = new Blob([csv], {type: "text/csv;charset=utf-8"});
+                    input_data_file = csv;
+                    excel_to_csv_ready = true;
```

Comment:
Instead of repeatedly calling `check` on a timeout to see if the Excel file has finished being converted, you should run the rest of the function code only after the `reader.onloadend` event has fired. This will probably require some refactoring (since you need to also run the rest of the code if the user uploads a regular CSV file), but will result in a better, more understandable structure.